seqr/views/apis/report_api_tests.py

```diff
@@ -582,39 +630,91 @@ def test_sample_metadata_export(self, mock_google_authenticated):
+            'The following columns are included in the "participant" data model but are missing in the report: ancestry_metadata',
+            'The following entries are missing recommended "phenotype_description" in the "participant" table: Broad_HG00731, Broad_HG00732, Broad_HG00733, Broad_NA20870, Broad_NA20872, Broad_NA20874, Broad_NA20875, Broad_NA20876, Broad_NA20877, Broad_NA20881',
+            'The following entries are missing recommended "age_at_enrollment" in the "participant" table: Broad_HG00731, Broad_HG00732, Broad_HG00733, Broad_NA19678, Broad_NA19679, Broad_NA20870, Broad_NA20872, Broad_NA20874, Broad_NA20875, Broad_NA20876, Broad_NA20877, Broad_NA20881',
+        ] + skipped_file_validation_warnings[1:6] + skipped_file_validation_warnings[7:])
+        self.assertListEqual(response.json()['errors'], [
+            'The following entries are missing required "proband_relationship" in the "participant" table: Broad_HG00731, Broad_HG00732, Broad_HG00733, Broad_NA19678, Broad_NA19679, Broad_NA20870, Broad_NA20872, Broad_NA20874, Broad_NA20875, Broad_NA20876, Broad_NA20877, Broad_NA20881',
+            'The following entries have invalid values for "reported_race" in the "participant" table. Allowed values: Asian, White, Black, Unknown. Invalid values: Broad_NA19675_1 (Middle Eastern or North African)',
+            'The following entries are missing required "aligned_dna_short_read_set_id" (from Airtable) in the "aligned_dna_short_read_set" table: NA19675_1',
+        ])
+
+        responses.add(responses.GET, MOCK_DATA_MODEL_URL, status=404)
+        responses.calls.reset()
+        mock_subprocess.reset_mock()
+        mock_google_authenticated.return_value = True
+        response = self.client.post(url, content_type='application/json', data=json.dumps(body))
         self.assertEqual(response.status_code, 200)
-        self.assertEqual(
-            response.get('content-disposition'),
-            'attachment; filename="GREGoR Reports HMB.zip"'
-        )
+        self.assertDictEqual(response.json(), {
+            'info': ['Successfully validated and uploaded Gregor Report for 9 families'],
+            'warnings': [
+                'Unable to load data model for validation: 404 Client Error: Not Found for url: http://raw.githubusercontent.com/gregor_data_model.json',
+            ] + skipped_file_validation_warnings,
+        })
 
-        files = self._get_zip_files(mock_zip, [
-            'participant.tsv', 'family.tsv', 'phenotype.tsv', 'analyte.tsv', 'experiment_dna_short_read.tsv',
-            'aligned_dna_short_read.tsv', 'aligned_dna_short_read_set.tsv', 'called_variants_dna_short_read.tsv',
-        ])
+        self.assertListEqual(
+            mock_open.call_args_list, [mock.call(f'/mock/tmp/{file}.tsv', 'w') for file in expected_files])
```

Comment:
It looks better than using the `assert_has_calls` function.

Reply:
In general I agree, but that doesn't work because the `assert_has_calls` also checks for all the calls to `__enter__` and `write` and it is really unreadable once those are all included (in addition to being like 32 calls). I really want to check that open is called the exact way we intend for it to, which is what this test does